[WPE][GTK] Speakers enumeration doesn't show ALSA virtual sinks

neodesys opened the pull request:
#### a420a8c5cf2113698903a874a565592cc18a127c
<pre>
[WPE][GTK] Speakers enumeration doesn&apos;t show ALSA virtual sinks
<a href="https://bugs.webkit.org/show_bug.cgi?id=300109">https://bugs.webkit.org/show_bug.cgi?id=300109</a>

Reviewed by Philippe Normand.

- The audio devices management system was considering the speakers as capture
  devices and so was automatically creating GStreamer pipelines with an
  underlying `alsasink` element opening the raw audio hardware. As ALSA
  devices have usually an exclusive access, maintaining the device open was
  impeding to list the other virtual sinks in the ALSA configuration. During
  the listing of the virtual sinks, it needs to open the underlying real
  audio device to fetch information about the audio samples format but
  it cannot because the audio real device is already open. In all cases, there
  should not be any GStreamer pipeline at all created for a speaker device as
  it produces no audio. Doing so was also producing some warnings at the
  GStreamer level as it was trying to set a property reserved for source
  elements on a sink element. The fix is filtering out the speaker devices
  when creating the `GStreamerAudioCaptureSource` instances in order to not
  build broken capture pipelines with sink elements.
- Indeed, the `GStreamerCaptureDeviceManager` was not creating unique IDs
  for the different input/output interfaces. Those unique IDs are used to link
  together different audio elements but, as they were not unique, a same sound
  card with a microphone and a speaker had the same ID for both interfaces. At
  the moment of creating a `GStreamerAudioCaptureSource` instance for the
  microphone, the system was fetching the speaker interface instead and not
  the microphone one and, so, was locking the real hardware sound card just
  like in the former point. In addition to the device name, the fix also adds
  the device GStreamer class to differentiate inputs from outputs in the
  device ID.

* Source/WebCore/Modules/speech/SpeechRecognitionCaptureSource.cpp:
(WebCore::SpeechRecognitionCaptureSource::findCaptureDevice):
* Source/WebCore/platform/mediastream/CaptureDevice.h:
(WebCore::CaptureDevice::isSpeakerDevice const):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::createMediaStream):
(WebCore::RealtimeMediaSourceCenter::getUserMediaDevices):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCaptureSource.cpp:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.cpp:
(WebCore::GStreamerCaptureDeviceManager::captureDeviceFromGstDevice):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCapturer.cpp:
(WebCore::GStreamerCapturer::createSource):
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:

Canonical link: <a href="https://commits.webkit.org/301082@main">https://commits.webkit.org/301082@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2b5902d30f85536f2ace7b26f02b5a296959f30b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124885 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44555 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35291 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131730 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126762 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45253 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53121 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95036 "22 flakes 26 failures") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127839 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36106 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111689 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75592 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35041 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/29845 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75209 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105866 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30074 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134400 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51722 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39523 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103515 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52141 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107904 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103289 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26294 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48638 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26928 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48743 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51599 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57406 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50986 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54342 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52679 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/mediastream/CaptureDevice.h:
```diff
@@ -93,6 +93,11 @@ class CaptureDevice {
         return m_type == DeviceType::Microphone || m_type == DeviceType::Camera;
     }
 
+    bool isOutputDevice() const
```
I'm not sure this is a good name, a window or screen device is also an output device... maybe make it more explicit, like `isSpeakerDevice()`.

neodesys replied:
`Window` and `Screen`, so as `SystemAudio`, in this context are used to capture the window, the screen content, or the audio monitor for sharing screen or window. This is why it is logical to have them in `CaptureDevice` because they are all used to capture audio and video streams.

So, `!isInputDevice()` is totally different from `isOutpuDevice()`. Maybe a better name for `isInputDevice()` would be more something like `isExternalDevice()` or something equivalent. Maybe the name might have been chosen as a reference to [MediaDeviceInfo::kind](https://developer.mozilla.org/en-US/docs/Web/API/MediaDeviceInfo/kind).

In reality, `Speaker` should not be part of `CaptureDevice` as it is not used to capture neither video, nor audio. But I suppose it has been implemented like this for convenience issues and to avoid to refactor all the code around.

As `Speaker` is the only type that needs to be filtered out, maybe I can just get rid of the `isOutputDevice()` function and directly compare the type to `Speaker` in all places where I call this function. Or just rename the function as you are suggesting.

What do you think ?

philn replied:
I'd go for the renamed method, but no objection if you prefer explicit type checks everywhere :)

neodesys replied:
:+1: Done.
I renamed the function `isSpeakerDevice`.

philn reviewed (commented):
Do we really need a new method in CaptureDevice? Why not use `!isInputDevice()`?

philn commented on Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
```diff
@@ -316,7 +316,7 @@ class MockRealtimeAudioSourceFactory final
     CaptureSourceOrError createAudioCaptureSource(const CaptureDevice& device, MediaDeviceHashSalts&& hashSalts, const MediaConstraints* constraints, std::optional<PageIdentifier> pageIdentifier) final
     {
-        ASSERT(device.type() == CaptureDevice::DeviceType::Microphone || device.type() == CaptureDevice::DeviceType::Speaker);
+        ASSERT(!device.isSpeakerDevice());
```
Hm why is this changed? The code below clearly implies speakers should be allowed.

neodesys replied:
I think I forgot to modify the error messages, but even for a mocked implementation `createAudioCaptureSource` should not return a `RealtimeSource` for speakers because it implies to build a source pipeline for fetching buffers from a source element which is incompatible with using a speaker device.

In fact I should have kept the `ASSERT(device.type() == CaptureDevice::DeviceType::Microphone);` only because originally the `MockRealtimeAudioSource::create()` called at the end of this function was only intended to create some `RealtimeSource` for microphones (https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/mediastream/gstreamer/MockRealtimeAudioSourceGStreamer.cpp#L64).

I suppose the `speaker` type has been added later in the assert to make tests passing also with speakers although they should not be associated with `RealtimeMediaSource`. Now that speakers have been filtered out from creating media sources, this assert is not needed anymore.

I've fixed it and also modified the error messages.

neodesys committed:
[WPE][GTK] Speakers enumeration doesn't show ALSA virtual sinks
https://bugs.webkit.org/show_bug.cgi?id=300109

Reviewed by Philippe Normand.

- The audio devices management system was considering the speakers as capture
  devices and so was automatically creating GStreamer pipelines with an
  underlying `alsasink` element opening the raw audio hardware. As ALSA
  devices have usually an exclusive access, maintaining the device open was
  impeding to list the other virtual sinks in the ALSA configuration. During
  the listing of the virtual sinks, it needs to open the underlying real
  audio device to fetch information about the audio samples format but
  it cannot because the audio real device is already open. In all cases, there
  should not be any GStreamer pipeline at all created for a speaker device as
  it produces no audio. Doing so was also producing some warnings at the
  GStreamer level as it was trying to set a property reserved for source
  elements on a sink element. The fix is filtering out the speaker devices
  when creating the `GStreamerAudioCaptureSource` instances in order to not
  build broken capture pipelines with sink elements.
- Indeed, the `GStreamerCaptureDeviceManager` was not creating unique IDs
  for the different input/output interfaces. Those unique IDs are used to link
  together different audio elements but, as they were not unique, a same sound
  card with a microphone and a speaker had the same ID for both interfaces. At
  the moment of creating a `GStreamerAudioCaptureSource` instance for the
  microphone, the system was fetching the speaker interface instead and not
  the microphone one and, so, was locking the real hardware sound card just
  like in the former point. In addition to the device name, the fix also adds
  the device GStreamer class to differentiate inputs from outputs in the
  device ID.

* Source/WebCore/Modules/speech/SpeechRecognitionCaptureSource.cpp:
(WebCore::SpeechRecognitionCaptureSource::findCaptureDevice):
* Source/WebCore/platform/mediastream/CaptureDevice.h:
(WebCore::CaptureDevice::isSpeakerDevice const):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::createMediaStream):
(WebCore::RealtimeMediaSourceCenter::getUserMediaDevices):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCaptureSource.cpp:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.cpp:
(WebCore::GStreamerCaptureDeviceManager::captureDeviceFromGstDevice):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCapturer.cpp:
(WebCore::GStreamerCapturer::createSource):
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:

Canonical link: https://commits.webkit.org/301082@main
